File: frontend/src/api/grouping.ts
import apiClient from './client'

export interface GroupingRunStartResponse {
  run_id: string
}

export interface GroupingStatusResponse {
  stage: string
  current: number
  total: number
  status: string
  groups_found?: number
  error?: string
}

export interface GroupingRunRead {
  id: string
  status: string
  threshold: number
  context: string
  started_at: string
  finished_at: string | null
  projects_processed: number | null
  groups_found: number | null
  projects_in_groups: number | null
  error_message: string | null
  confirmed_rate: number | null
}

export interface GroupingHistoryResponse {
  items: GroupingRunRead[]
  total: number
}

export const groupingApi = {
  async startGrouping(threshold: number, context: string): Promise<GroupingRunStartResponse> {
    const { data } = await apiClient.post<GroupingRunStartResponse>('/grouping/run', {
      threshold,
      context,
    })
    return data
  },

  async getGroupingStatus(runId: string): Promise<GroupingStatusResponse> {
    const { data } = await apiClient.get<GroupingStatusResponse>(`/grouping/status/${runId}`)
    return data
  },

  async getGroupingHistory(): Promise<GroupingHistoryResponse> {
    const { data } = await apiClient.get<GroupingHistoryResponse>('/grouping/history')
    return data
  },
}

File: backend/app/schemas/grouping.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models import GroupContext, RunStatus


class GroupingRunCreate(BaseModel):
    threshold: float
    context: GroupContext = GroupContext.main


class GroupingRunRead(BaseModel):
    id: uuid.UUID
    status: RunStatus
    threshold: float
    context: GroupContext
    started_at: datetime
    finished_at: datetime | None = None
    projects_processed: int | None = None
    groups_found: int | None = None
    projects_in_groups: int | None = None
    error_message: str | None = None
    confirmed_rate: float | None = None

    model_config = {"from_attributes": True}


class GroupingStatusResponse(BaseModel):
    run_id: str
    stage: str | None = None
    current: int | None = None
    total: int | None = None
    status: str | None = None
    error_message: str | None = None
    groups_found: int | None = None


class GroupingRunStartRequest(BaseModel):
    threshold: float = Field(default=0.75, ge=0.5, le=0.95)
    context: str = "main"


class GroupingRunStartResponse(BaseModel):
    run_id: str


class GroupingHistoryResponse(BaseModel):
    items: list[GroupingRunRead]
    total: int
